### ISSUE TITLE
Refinement Region problems

### ISSUE THREAD

**Issue by gamemakerh** [ladybug-tools/butterfly#290]
Hi for everyone,
I encountered problem and wanna your guidance/opinion/help...

I want to refine specific region. In my case there is intersection volume between domain, cone and cylinder. 
Please check picture below.
![image](https://user-images.githubusercontent.com/4093017/32533035-efb5c5f4-c456-11e7-8cb7-858bb90e4bfb.png)

Using "**Distance Refinement Mode**" block everything seems to work fine
![image](https://user-images.githubusercontent.com/4093017/32533099-4995121e-c457-11e7-8807-993a9ab5d6b9.png)
(geo is cone)
![image](https://user-images.githubusercontent.com/4093017/32533188-f36acf90-c457-11e7-8242-7331002788f4.png)

But using **In Out Refinement Mode** I get an error every time I try to implement.

```


--> FOAM FATAL IO ERROR: 
wrong token type - expected int32_t, found on line 163 the word 'L'

file: /home/ofuser/workingDir/butterfly/Duct_Simulation430oclock_5/system/snappyHexMeshDict.castellatedMeshControls.refinementRegions.refinement1.levels at line 163.

    From function Foam::Istream& Foam::operator>>(Foam::Istream&, int32_t&)
    in file primitives/ints/int32/int32IO.C at line 67.

FOAM exiting


```

Any ideas  what I'm doing wrong ?

**Comment by gamemakerh** [comment]
ok one more thing. this is bug or what ?

from Openfoma Mesh Refinement Documentary you can find that Region refinement can be described as Float number, but on on Butterfly. I'm afraid this is bug. Butterfly supports only Integers ?
![image](https://user-images.githubusercontent.com/4093017/32536528-85af10be-c467-11e7-9316-a22baa165f4c.png)

```
Region-wise refinement

Section for refinement's selection of user defined regions by three methods explained in sHMDict comments.
Sometimes different type regions tend to override each other, introduce them carefully.

// Region-wise refinement
// ~~~~~~~~~~~~~~~~~~~~~~
// Specifies refinement level for cells in relation to a surface. One of
// three modes
// - distance. 'levels' specifies per distance to the surface the
// wanted refinement level. The distances need to be specified in
// descending order.
// - inside. 'levels' is only one entry and only the level is used. All
// cells inside the surface get refined up to the level. The surface
// needs to be closed for this to be possible.
// - outside. Same but cells outside.

refinementRegions
{
        refinementBox
        {
            mode distance;
            levels ((0.05 5) (0.1 4));
        }
        refinementSphere
        {
            mode inside;
            levels ((0.05 4));
        }
        refinementCylinder
        {
            mode outside;
            levels ((0.05 4));
        }
}

```

**Comment by mostaphaRoudsari** [comment]
@gamemakerh, thanks for posting both and sorry for the very late reply.

Back to `wrong token type - expected int32_t, found on line 163 the word 'L'`:
I believe this is discussed a different issue #279 and [is fixed now](https://github.com/ladybug-tools/butterfly/commit/0b18f94d8faa983129d63f2cfa304fd913741739).

>> from Openfoma Mesh Refinement Documentary you can find that Region refinement can be described as Float number, but on on Butterfly. I'm afraid this is bug. Butterfly supports only Integers ?

This looks like a bug but is easy to fix. If I remember correctly it was used to be float but then we had issues in some cases and set it to integer but based on your example it should be the other way around. Did you find a workaround for this case?

**Comment by gamemakerh** [comment]
Ok I've re updated butterfly component, but I'm still getting error message from snappyHexMesh block.
![image](https://user-images.githubusercontent.com/4093017/33173171-c0635706-d05b-11e7-946c-2ca9f045f0a1.png)

```
SnappyHexMeshDict.addLayersControls.minThickness is changed from "0.2" to "0.1".
SnappyHexMeshDict.addLayersControls.castellatedMeshControls.nCellsBetweenLevels is changed from "3" to "10".
saving the new snappyHexMeshDict.
docker-machine env
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
DOCKER_MACHINE_NAME=default&SET COMPOSE_CONVERT_WINDOWS_PATHS=true&docker exec -i 2c11949b2258 pgrep snappyHexMesh
docker-machine env
Butterfly is running snappyHexMesh. PID: None
/*---------------------------------------------------------------------------*\
| =========                 |                                                 |
| \\      /  F ield         | OpenFOAM: The Open Source CFD Toolbox           |
|  \\    /   O peration     | Version:  v1706                                 |
|   \\  /    A nd           | Web:      www.OpenFOAM.com                      |
|    \\/     M anipulation  |                                                 |
\*---------------------------------------------------------------------------*/
Build  : v1706
Arch   : "LSB;label=32;scalar=64"
Exec   : snappyHexMesh
Date   : Nov 23 2017
Time   : 12:35:17
Host   : "default"
PID    : 5776
Case   : /home/ofuser/workingDir/butterfly/Duct_Simulation430oclock_5
nProcs : 1
sigFpe : Enabling floating point exception trapping (FOAM_SIGFPE).
fileModificationChecking : Monitoring run-time modified files using timeStampMaster (fileModificationSkew 10)
allowSystemOperations : Allowing user-supplied system call operations

// * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * //
Create time

Create mesh for time = 0

Read mesh in = 0.03 s

Overall mesh bounding box  : (-11.5138 -10.2438 0.45) (9.9888 3.4818 0.55)
Relative tolerance         : 1e-06
Absolute matching distance : 2.551007e-05

Reading refinement surfaces.
Read refinement surfaces in = 0.01 s

Reading refinement shells.
--> FOAM FATAL ERROR: 
Shell Refinement_1_Cone_Cross_Section.stl does not support testing for inside
Probably it is not closed.

    From function void Foam::shellSurfaces::setAndCheckLevels(Foam::label, const Foam::List<Foam::Tuple2<double, int> >&)
    in file shellSurfaces/shellSurfaces.C at line 123.

FOAM exiting
Runtime error (PythonException): 
 --> snappyHexMesh Failed!
--> FOAM FATAL ERROR: 
Shell Refinement_1_Cone_Cross_Section.stl does not support testing for inside
Probably it is not closed.

    From function void Foam::shellSurfaces::setAndCheckLevels(Foam::label, const Foam::List<Foam::Tuple2<double, int> >&)
    in file shellSurfaces/shellSurfaces.C at line 123.

FOAM exiting
Traceback:
  line 110, in script

```

**Comment by mostaphaRoudsari** [comment]
> Shell Refinement_1_Cone_Cross_Section.stl does not support testing for inside. Probably it is not closed.

You can use *inside/outside mode* only for closed geometries. In this case you should use distance.

The values are set to integer in [this line](https://github.com/ladybug-tools/butterfly/blob/0b18f94d8faa983129d63f2cfa304fd913741739/butterfly/refinementRegion.py#L72) but changing it to float will change the type to float for other refinement classes. I'm fixing this issue in the pep8 branch.

**Comment by mostaphaRoudsari** [comment]
I have to do some test before changing the types. Here is why we changed them to integers. https://github.com/ladybug-tools/butterfly/issues/257

**Comment by mostaphaRoudsari** [comment]
I think now it's clear. The first value can be a float, the second value must be an integer.

```cpp
refinementRegions
  {
      box1x1x1
      {
          mode inside;
          levels ((1.0 4));         // refinement level 4 (1.0 entry ignored)
      }
  
      sphere.stl
      {                             // refinement level 5 within 1.0 m
          mode distance;            // refinement level 3 within 2.0 m
          levels ((1.0 5) (2.0 3)); // levels must be ordered nearest first
      }
  }
```

**Comment by gamemakerh** [comment]
Thank you @mostaphaRoudsari   !

**Comment by gamemakerh** [comment]
@mostaphaRoudsari  sorry for this question but I'm really have no idea how to "update" my BF to have this adjustment of the distanceRefMode. Every time I get an error 

> 1. Solution exception:
> Failed to import butterfly:
> Cannot import name bf_geometry_from_stl_file


![image](https://user-images.githubusercontent.com/4093017/33200910-24803ae4-d0ff-11e7-8b66-2ce5fb4a446e.png)


I've tried to launch  **butterfly-pep8**
but many errors  appeared on my GH case. most of the BF blocks was "RED"

then I've switched back to the "Original" BF and changed **refinementRegion.py** file.
Then I've got refinement region blocks "RED"

-------------

Ref regions blocks working with PEP8 version, but I need all other blocks :) 

There is somehow solution to do it right ? have floats on ref regions and have all the rest of the model working ?
